``` {r setup, echo=FALSE, message=FALSE, include=FALSE, error=FALSE}
library(dplyr)
library(magrittr)
library(BSgenome.Dmelanogaster.UCSC.dm3)
library(ggplot2)
library(stringr)

# Output folder for this document
options(knitr.figure_dir = "figure_3a_4b_cbp_enhancers_motif_enrichment")

set.seed(12345)

source("shared_code/knitr_common.r")
source("shared_code/granges_common.r")
source("shared_code/flybase.r")
source("shared_code/motif_enrichments.r")
source("shared_code/motif_conservation.r")
source("shared_code/motif_overlaps.r")
source("shared_code/seq_logo.r")
source("shared_code/rnaseq.r")
source("shared_code/ggplot_theme.r")
source("shared_code/stat_tests.r")
```

# Figure 3a and 4b: CBP enhancers known motif enrichment analysis

``` {r header_child, child="child_docs/header_child.Rmd"}
```

``` {r data_paths, include=FALSE}

motif_granges_path <- function() {
 data_path("fimo/granges")
}

source_peaks <- function() {
  readRDS("figure_1c_cbp_peaks_with_differential_h3k27ac/final_peaks.gr.rds") %>%
    resize(width=201, fix="center") %>%
    subset(classification != "TSS")
}

pv_cutoff <- function() {
  return(0.05)
}

cores <- function() {
  return(7)
}

transcription_factor_fpkm_expression_threshold <- 3

```

## Region counts

``` {r region_counts_table}

source_peaks() %>%
  as.data.frame %>%
  group_by(diff_bound_class) %>%
  summarize(count = n()) %>%
  pander

```

## Assign FlyBase gene IDs to known motifs

``` {r match_motifs}
motif_ids.df <- match_motifs_to_fb_genes(list.files(motif_granges_path(), ".granges.rds$"))
```

## Calculate enrichments

We can now test for enrichment of known motifs using the results from DESeq. We will use an adjusted p-value cutoff of `r pv_cutoff()`. Each Nejire peak has been resized to 201 bp centered on the Nejire ChIP-seq summit.

We do not test the following motifs:

  - All JASPAR\_CORE motifs, as they are duplicates of JASPAR\_2014
  - All daughterless (da) hetereodimer motifs
  - All motifs where the corresponding transcription factor is not expressed in any of the three tissues (FPKM > 5)

``` {r motif_enrichments, include=FALSE}

motif_results <- function(peaks.gr, motifs) {
  f.gd7 <- subset(peaks.gr, diff_bound_class == "Higher in gd7")
  f.tol <- subset(peaks.gr, diff_bound_class == "Higher in Toll10b")
  f.nc  <- subset(peaks.gr, diff_bound_class == "No change")

  counts.gd7 <- motif_counts_rds(f.gd7, motifs, cores=cores())
  counts.tol <- motif_counts_rds(f.tol, motifs, cores=cores())
  counts.nc  <- motif_counts_rds(f.nc,  motifs, cores=cores())

  gd7.results <- motif_count_comparison(counts.gd7, counts.nc)
  gd7.results$diff_bound_class <- "Higher in gd7"
  
  tol.results <- motif_count_comparison(counts.tol, counts.nc)
  tol.results$diff_bound_class <- "Higher in Toll10b"

  direct.results <- motif_count_comparison(counts.gd7, counts.tol)
  direct.results$diff_bound_class <- "gd7 vs Toll10b"
  
  results.df <- bind_rows(gd7.results, tol.results, direct.results)
  results.df
}

motifs <- list.files(motif_granges_path(), "\\.granges\\.rds$", full.names=TRUE)
names(motifs) <- gsub("\\.granges\\.rds$", "", basename(motifs))

expressed_genes <- genes_expressed_in_tissue(c("gd", "toll"), fpkm_threshold=transcription_factor_fpkm_expression_threshold)

test_motifs <- subset(motif_ids.df, motif %in% names(motifs) & motif_fb_id %in% expressed_genes)$motif

motifs <- motifs[names(motifs) %in% test_motifs]
motifs <- motifs[-grep("^JASPAR_CORE", names(motifs))]
motifs <- motifs[-grep("_da_", names(motifs))]
motifs <- motifs[-grep("tgo_", names(motifs))]

motif_results.df <- cache("motif_results.df", function() {

  peaks.gr <- source_peaks()
  
  results <- motif_results(peaks.gr, motifs)
  results
})

```

This results in `r length(motifs)` tested motifs.

``` {r process_motif_enrichments, include=FALSE}

preprocess_motif_results <- function(motif_results.all) {
  motif_results.all$enrichment[is.infinite(motif_results.all$enrichment)] <- 1
  motif_results.all$enrichment[motif_results.all$enrichment == 0] <- 1
  motif_results.all$enrichment[is.na(motif_results.all$enrichment)] <- 1
  motif_results.all
}

results.df <- preprocess_motif_results(motif_results.df)

results.df$group_label <- results.df$diff_bound_class
results.df <- merge(results.df, motif_ids.df, all.x=TRUE)

results.df$adj_pv <- p.adjust(results.df$pvalue, method="BH")

saveRDS(results.df, file=figure_path("filtered_results.df.rds"))
```

After removal, we calculate an adjusted p-value for each motif enrichment test.

## Calculate overlap values for pairs of significant motifs

``` {r calculate_overlaps, message=FALSE}

sig_motifs <- c(subset(results.df, adj_pv < pv_cutoff() & diff_bound_class == "gd7 vs Toll10b")$motif,
                subset(results.df, adj_pv < pv_cutoff() & test_type == "enrichment")$motif) %>% unique

differential_peaks.gr <- subset(source_peaks(), diff_bound_class != "No change")

overlaps.df <- cache("overlaps.df", function() {
  overlap_table_for_motifs(sig_motifs, differential_peaks.gr)
})

```

We next calculate the overlap between all pairs of `r length(sig_motifs)` significant motifs.

## Enrichment and overlap heat maps

``` {r build_heatmaps_and_display_overlap_tables, message=FALSE}

motif_figure_order <- c("dl",
                        "CG16778",
                        "Tin",
                        "rib",
                        "vfl",
                        "sna",
                        "eve",
                        "Dll",
                        "BEAF",
                        "HLH",
                        "AbdB",
                        "brk",
                        "kni")

overlap_percent <- 10

overlaps.df$overlap <- pmax(overlaps.df$m1_percent, overlaps.df$m2_percent)

include_motifs <- sig_motifs #subset(results.df, adj_pv < pv_cutoff())$motif
sig.df <- subset(results.df, motif %in% include_motifs)

sig.overlap <- group_by_overlap(sig.df, overlaps.df, percent=overlap_percent)

sig.overlap$df$other_names %<>% gsub(", ", "<br>", .)
sig.overlap$df %>% pander(caption="Motif overlap table")

saveRDS(sig.overlap, file=figure_path("motifs_grouped_by_overlap.df.rds"))

sig.df <- sig.overlap$filtered

sig_motifs <- sig.df$motif %>% unique
motifs_ordered <- lapply(motif_figure_order, function(m) {
            sig_motifs[grep(m, sig_motifs)[1]]
          }) %>%
          unlist

stopifnot(length(motifs_ordered) == length(motif_figure_order))

g <- clustered_motif_heatmap(sig.df, "Differential H3K27ac at Nej", motif.order=rev(motifs_ordered))
```

Motif pairs with more than `r overlap_percent`% overlapping instances are combined, selecting the motif with the best p-value.

``` {r show_motif_enrichment_heatmaps_inline, eval=TRUE, fig.cap="", fig.width=8, fig.height=8, dev=c("png", "pdf")}
g
```

``` {r show_motif_overlap_heatmaps_inline, eval=TRUE, fig.cap="", fig.width=14, fig.height=10, dev=c("png", "pdf")}

overlap_heatmap.df <- subset(overlaps.df, motif1 %in% sig.df$motif & motif2 %in% sig.df$motif)
overlap_heatmap.df$motif1 %<>% factor(levels=levels(g$data$motif))
overlap_heatmap.df$motif2 %<>% factor(levels=levels(g$data$motif))

overlap_heatmap.df2 <- overlap_heatmap.df
names(overlap_heatmap.df2)[1:2] <- c("motif2", "motif1")
overlap_heatmap.df %<>% bind_rows(overlap_heatmap.df2)

g_overlaps <- ggplot(overlap_heatmap.df, aes(x=motif1, y=motif2, fill=overlap)) +
              geom_tile() +
              geom_text(aes(label=as.integer(overlap)), color="black") +
              scale_y_discrete(expand=c(0, 0)) +
              scale_x_discrete(expand=c(0, 0)) +
              scale_fill_gradient(name="Overlap%", high="red", low="white", limits=c(0, 100)) +
              labs(x="", y="", title="Overlapping motif instances") +
              theme_bw() +
              theme(axis.text.x=element_text(angle=45, hjust=1, vjust=1),
                    axis.text.y=element_text(hjust=1),
                    panel.grid.minor=element_blank(),
                    panel.grid.major=element_blank(),
                    axis.line=element_blank(),
                    axis.ticks=element_line(size = 0.5, color="gray50"))
              
g_overlaps
```

## Motif conservation

``` {r motif_conservation, fig.cap="", fig.width=16, fig.height=7}

phastcon.bw <- data_path("phastcon/phastcon_15way.bw")

phastcon.df <- cache("phastcon.df", function() {
  sig.df$motif %>%
    unique %>%
    mclapply(function(motif) {
      toll.df <- phastcon_scores_for_motif_instances(motif, subset(differential_peaks.gr, diff_bound_class == "Higher in Toll10b"), phastcon.bw)
      toll.df$enhancer_group <- "Higher in Toll10b"
   
      gd7.df <- phastcon_scores_for_motif_instances(motif, subset(differential_peaks.gr, diff_bound_class == "Higher in gd7"), phastcon.bw)
      gd7.df$enhancer_group <- "Higher in gd7"
   
      rbind(toll.df, gd7.df)
    }, mc.cores=cores()) %>%
    bind_rows
})

# Add entire enhancer scores
phastcon_enhancers_toll.df <- data_frame(motif=" Entire enhancer", 
                                         phastcon_score=regionMeans(subset(differential_peaks.gr, diff_bound_class == "Higher in Toll10b"), phastcon.bw),
                                         enhancer_group="Higher in Toll10b")

phastcon_enhancers_gd.df <- data_frame(motif=" Entire enhancer", 
                                       phastcon_score=regionMeans(subset(differential_peaks.gr, diff_bound_class == "Higher in gd7"), phastcon.bw),
                                       enhancer_group="Higher in gd7")

phastcon.df %<>% rbind(phastcon_enhancers_toll.df, phastcon_enhancers_gd.df)

phastcon.df$motif %<>% gsub("FlyFactorSurvey-", "FFS-", .)

all_enhancers_median_phastcon_score <- median(c(phastcon_enhancers_toll.df$phastcon_score,
                                                phastcon_enhancers_gd.df$phastcon_score))



g <- ggplot(phastcon.df, aes(x=motif, y=phastcon_score, fill=enhancer_group)) +
     geom_hline(yintercept=all_enhancers_median_phastcon_score, color="blue", linetype="dotted") +
     geom_boxplot() +
     scale_fill_discrete("H3K27Ac status") +
     labs(x="", y="PhastCon score", title="PhastCon scores of motif instances") +
     theme_bw() +
     theme(axis.text.x=element_text(angle=45, hjust=1, vjust=1))
#g
```

``` {r motif_conservation_heatmap, fig.cap="", fig.width=7, fig.height=8, dev=c("png", "pdf")}

cons.df <- cache("cons.df", function() {
  sig.df$motif %>%
    unique %>%
    checked_mclapply(function(motif) {
      phastcon_scores_for_motif_instances_vs_genome(motif, differential_peaks.gr, phastcon.bw)
    }, mc.cores=cores(), mc.preschedule=FALSE) %>%
    bind_rows
})

summary.df <- cons.df %>%
              group_by(motif) %>%
              summarize(WithinEnhancers = mean(phastcon_score[group == "region"]),
                        Genome = mean(phastcon_score[group == "genome"])) %>%
              ungroup %>%
              gather(motif_group, score, WithinEnhancers:Genome)

summary.df$motif_group <- factor(summary.df$motif_group, levels=c("Genome", "WithinEnhancers"))
#summary.df$motif <- factor(summary.df$motif, levels=rev(motifs))

g <- ggplot(summary.df, aes(x=motif_group, y=motif, fill=score * 100)) +
     geom_tile() +
     theme_bw() +
     scale_x_discrete(expand=c(0, 0)) +
     scale_y_discrete(expand=c(0, 0)) +
     theme(panel.grid.major=element_blank(),
           panel.grid.minor=element_blank()) +
     labs(x="", y="") +
     scale_fill_gradientn(name="Conservation", space="Lab",
                          values=c(30, 100), 
                          colours=c("white", "darkred"), 
                          rescaler=function(x,...) x, oob=identity,
                          limits=c(min(c(30, summary.df$score*100)), 100),
                          guide=guide_colorbar())
#g
```

An alternative conservation plot:

  - Background control: random non-TSS regions (201 bp)
  - Average phastCon score of enhancers with each motif
  - Average phastCon score of each motif

``` {r motif_conservation_barplots, fig.cap="", fig.width=13, fig.height=6, dev=c("png", "pdf")}

reduce_motif <- function(gr) {
  strand(gr) <- "+"
  reduce(gr)
}

random.gr <- cache("random.gr", function() {
  gr <- tileGenome(seqlengths=seqlengths(Dmelanogaster), tilewidth=201, cut.last=TRUE) %>%
        filter_chrs %>%
        assign_nearest_gene %>%
        subset(distance_to_tss > 2000) %>%
        sample(5000)
  gr$phastcon_score <- regionMeans(gr, phastcon.bw)
  gr
})

differential_peaks.gr$phastcon_score <- regionMeans(differential_peaks.gr, phastcon.bw)

with_motif.grl <- sig_motifs %>%
                  checked_mclapply(function(motif_name) {
                    motifs.gr <- load_motif(motif_name)
                    ol <- findOverlaps(query=motifs.gr, subject=differential_peaks.gr, type="within", ignore.strand=TRUE)
                    differential_peaks.gr[unique(subjectHits(ol))]
                  })
names(with_motif.grl) <- sig_motifs

barplots.df <- sig_motifs %>%
               lapply(function(motif_name) {
                 enhancers <- data.frame(stringsAsFactors=FALSE,
                                         n = length(with_motif.grl[[motif_name]]),
                                         bargroup = "Enhancers with motif",
                                         phastcon_score = mean(with_motif.grl[[motif_name]]$phastcon_score))
                 motifs <- data.frame(stringsAsFactors=FALSE,
                                      n = nrow(subset(cons.df, group == "region" & motif == motif_name)),
                                      bargroup = "Motif instances",
                                      phastcon_score = mean(subset(cons.df, group == "region" & motif == motif_name)$phastcon_score))

                 combined.df <- bind_rows(enhancers, motifs)
                 combined.df$motif <- motif_name
                 combined.df
               }) %>%
               bind_rows

control.df <- data.frame(stringsAsFactors=FALSE,
                         motif = "Control",
                         bargroup = c("Random non-TSS regions", "All enhancers"),
                         phastcon_score = c(mean(random.gr$phastcon_score), mean(differential_peaks.gr$phastcon_score)),
                         n = c(length(random.gr), length(differential_peaks.gr)),
                         pvalue_all_vs_random = wilcox.test(x=differential_peaks.gr$phastcon_score, 
                                                            y=random.gr$phastcon_score,
                                                            alternative="greater", paired=FALSE)$p.value)


cons_pairs.df <- names(with_motif.grl) %>%
                 lapply(function(motif_name) {
                   motifs.gr <- load_motif(motif_name) %>% reduce_motif
                   enhancers.gr <- with_motif.grl[[motif_name]]
              
                   ol <- findOverlaps(query=motifs.gr, subject=enhancers.gr, type="within", ignore.strand=TRUE)
              
                   motifs_inside.gr <- motifs.gr[unique(queryHits(ol))]
                   motifs_inside.gr$phastcon_score <- regionMeans(motifs_inside.gr, phastcon.bw)
              
                   ol <- findOverlaps(query=motifs_inside.gr, subject=enhancers.gr, type="within", ignore.strand=TRUE) %>% as.data.frame
              
                   ol$enhancer_score <- enhancers.gr$phastcon_score[ol$subjectHits]
                   ol$enhancer_name <- enhancers.gr$name[ol$subjectHits]
                   ol$motif_score <- motifs_inside.gr$phastcon_score[ol$queryHits]
                   ol$motif_width <- width(motifs_inside.gr)[ol$queryHits]

                   per_enhancer <- ol %>%
                                   group_by(subjectHits) %>%
                                   summarize(enhancer_name = enhancer_name[1],
                                             enhancer_score = enhancer_score[1],
                                             motif_score = weighted.mean(motif_score, motif_width),
                                             motif = motif_name)
                   per_enhancer
                 }) %>%
                 bind_rows


cons_pvalues.df <- cons_pairs.df %>%
                   group_by(motif) %>%
                   summarize(enhancer_count = n(),
                             mean_motif = mean(motif_score),
                             mean_enhancer = mean(enhancer_score),
                             pvalue = wilcox.test(x = motif_score, y = enhancer_score, paired=TRUE, alternative="greater")$p.value)


barplots.df %<>% bind_rows(select(control.df, -pvalue_all_vs_random))
barplots.df$motif <- gsub("FlyFactorSurvey", "FF", barplots.df$motif)
cons_pvalues.df$motif <- gsub("FlyFactorSurvey", "FF", cons_pvalues.df$motif)

barplot_pvalues.df <- barplots.df %>%
                      filter(bargroup %in% c("All enhancers", "Motif instances")) %>%
                      merge(cons_pvalues.df) %>%
                      mutate(significance = ifelse(pvalue < 0.05, "     *", ""))

barplots_motifs <- barplots.df$motif %>% unique
barplots_motifs <- c("Control", motif_figure_order) %>%
                   lapply(function(m) {
                     barplots_motifs[grep(m, barplots_motifs)]
                   })

barplots.df$motif %<>% factor(levels=barplots_motifs)

bargroup_colors <- c("All enhancers"="#CEAD8A",
                     "Random non-TSS regions"="#A27C63",
                     "Enhancers with motif"="#ABDBE9",
                     "Motif instances"="#337691")

g <- ggplot(barplots.df, aes(x=motif, y=phastcon_score * 100, fill=bargroup)) +
     geom_bar(stat="identity", position="dodge") +
     geom_text(data=barplot_pvalues.df, aes(label=significance), size=5, vjust=-0.5) +
     labs(x="", y="Average phastCon score") +
     scale_fill_manual("", values=bargroup_colors) +
     scale_y_continuous(expand=c(0, 0), limits=c(0, 100)) +
     theme_bw() +
     theme(axis.text.x=element_text(angle=45, hjust=1, vjust=1),
           panel.grid.major=element_blank(),
           panel.grid.minor=element_blank()) 

g



cons_boxplot.df <- cons_pairs.df %>%
                   mutate(score = enhancer_score, 
                          bargroup = "Enhancers with motif") %>%
                   select(motif, score, bargroup) %>%
                   rbind(cons_pairs.df %>%
                         mutate(score = motif_score,
                                bargroup = "Motif instances within enhancers") %>%
                         select(motif, score, bargroup))

controls_random.df <- data_frame(motif = "Control",
                                 bargroup = "Random non-TSS regions",
                                 score = random.gr$phastcon_score)
controls_all.df    <- data_frame(motif = "Control",
                                 bargroup = "All enhancers",
                                 score = differential_peaks.gr$phastcon_score)

boxplot.df <- bind_rows(cons_boxplot.df, controls_random.df, controls_all.df)
boxplot.df$motif <- gsub("FlyFactorSurvey", "FF", boxplot.df$motif)

g <- ggplot(boxplot.df, aes(x=motif, y=score, fill=bargroup)) +
     geom_boxplot(outlier.colour="gray50", outlier.size=1) +
     scale_fill_discrete("") +
     labs(x="", y="Average phastCon score") +
     theme_bw() +
     theme(axis.text.x=element_text(angle=45, hjust=1, vjust=1),
           panel.grid.major=element_blank(),
           panel.grid.minor=element_blank()) 
#g
```

Significance testing for "Enhancers with motif" and "Motif instances":
 
  - Wilcox paired rank-sum test for the average conservation of all motif instances in each enhancer compared to the average conservation of the enhancer itself
  
``` {r motif_conservation_pvalues}
control.df %>% pander("All enhancers versus random non-TSS regions (unpaired)")
cons_pvalues.df %>% pander("Motif instances versus enhancers with motif (paired)")
```

## Motif co-occurrence

``` {r motif_cooccurrence, message=FALSE}



compare_motifs <- function(peaks.gr, motif1.gr, motif2.gr) {
  
  peaks.gr$name <- paste0("peak_", seq_along(peaks.gr))
  
  motif1_ids <- mcols(peaks.gr)$name[countOverlaps(peaks.gr, motif1.gr, ignore.strand=TRUE) > 0]
  motif2_ids <- mcols(peaks.gr)$name[countOverlaps(peaks.gr, motif2.gr, ignore.strand=TRUE) > 0]

  universe_ids <- mcols(peaks.gr)$name

  test.df <- fisher_test_2x2(motif1_ids, motif2_ids, universe_ids, verbose=FALSE)
  test.df$motif1 <- motif1.gr$motif[1]
  test.df$motif2 <- motif2.gr$motif[1]
  test.df
}

combo_loop <- function(i, combos.df, peaks.gr, motifs.grl) {
  compare_motifs(peaks.gr, motifs.grl[[combos.df$motif1[i]]], motifs.grl[[combos.df$motif2[i]]])
}

motifs_grouped <- sig.overlap$flat

motifs.grl <- sig_motifs %>%
              checked_mclapply(function(motif) {
                all_motifs <- c(motif, subset(motifs_grouped, keep_motif == motif)$child_motif) %>%
                              unique %>%
                              lapply(load_motif) %>%
                              do.call(c, .) %>%
                              reduce_motif
                all_motifs$motif <- motif
                all_motifs
              }, mc.cores=cores(), mc.preschedule=FALSE)
names(motifs.grl) <- sig_motifs

pairs.df <- t(combn(sig_motifs, 2)) %>%
            as.data.frame(stringsAsFactors=FALSE)
names(pairs.df) <- c("motif1", "motif2")

coo.df <- 1:nrow(pairs.df) %>%
          checked_mclapply(combo_loop, pairs.df, source_peaks(), motifs.grl, mc.cores=cores(), mc.preschedule=FALSE) %>%
          bind_rows

coo.df$adj_pv <- p.adjust(coo.df$pvalue, method="BH")

coo.df %>%
  subset(adj_pv < 0.05) %>%
  pander(caption="Significant motif co-occurrence among all non-TSS Nejire peaks")

```

## Individual motifs

A PDF of motif logos for all significant motifs has been generated in the output folder for this document.

``` {r show_motifs, eval=TRUE, include=FALSE}

print_genome_motif <- function(motif_name) {
  motifs.gr <- load_motif(motif_name)

  if(length(motifs.gr) > 5000) motifs.gr <- sample(motifs.gr, 5000)
  motifs.seq <- getSeq(Dmelanogaster, motifs.gr)
  m <- consensusMatrix(motifs.seq)[1:4, ]
  grid.newpage()
  seqLogoGrid2(m / colSums(m), title=paste0("Genome: ", motif_name))
}

print_observed_motif <- function(motif_name, peaks.gr) {
  motifs.gr <- load_motif(motif_name)

  ol <- findOverlaps(motifs.gr, peaks.gr, type="within", ignore.strand=TRUE)
  motifs.seq <- getSeq(Dmelanogaster, motifs.gr[unique(queryHits(ol))])
  m <- consensusMatrix(motifs.seq)[1:4, ]
  grid.newpage()
  seqLogoGrid2(m / colSums(m), title=paste0("Observed: ", motif_name))
}

motifs <- unique(include_motifs)

pdf(figure_path("significant_motif_logos.pdf"), width=6, height=4)
nothing <- lapply(motifs, print_genome_motif)
dev.off()

```

``` {r session_info_child, child="child_docs/session_info_child.Rmd"}
```
